src/portable/raspberrypi/rp2040/rp2040_usb.c

```diff
@@ -126,8 +126,30 @@ void _hw_endpoint_start_next_buffer(struct hw_endpoint *ep)
 
     // PID
     val |= ep->next_pid ? USB_BUF_CTRL_DATA1_PID : USB_BUF_CTRL_DATA0_PID;
+
+#if TUSB_OPT_DEVICE_ENABLED
     ep->next_pid ^= 1u;
 
+#else
+    // For Host (also device but since we dictate the endpoint size, following scenario does not occur)
+    // Next PID depends on the number of packet in case wMaxPacketSize < 64 (e.g Interrupt Endpoint 8, or 12)
+    // Special case with control status stage where PID is always DATA1
+    if ( ep->transfer_size == 0 )
+    {
+      // ZLP also toggle data
+      ep->next_pid ^= 1u;
+    }else
+    {
+      uint32_t packet_count = 1 + ((ep->transfer_size - 1) / ep->wMaxPacketSize);
+
+      if ( packet_count & 0x01 )
```

Comment:
I guess to be clear, this was the line i was actually complaining about (based on the above).... for the the packets where packet_count is 8 the pid won't toggle, for all the ones packet count is 5 it will.

Reply:
I see what you mean there, you suspect that all the 8 packets of first 64 bytes has the same PID right ?

Reply:
correct

Reply:
Looking at the trace above the DATA PID in low speed mode needs to be toggled every 8 bytes. It depends on the approach. You can either limit the ep->transfer_size to do several 8 byte transfers. OR you can start a 64 byte transfer in low speed mode and the hardware will flip the PID for you every 8 bytes.

Reply:
> correct

That isn't the case here, the controller is very smart to toggle the PID every packet as shown in the captured which is actual capture to develop this PR. We only need to write the correct initial PID. 

> You can either limit the ep->transfer_size to do several 8 byte transfers. OR you can start a 64 byte transfer in low speed mode and the hardware will flip the PID for you every 8 bytes.

I tried both way, both works. Actually the code used to limit transfer_size at 8 (max packet size), but having it with as 64 make it much more faster, this will also help to enable the double buffering later on as well.

Reply:
@kilograham do you agree this is resolved if the hardware toggles the pid every 8 bytes in LS mode?

Reply:
actually this is only tested with IN transfer, therefore the PID is toggled from the device side. I haven't tested the OUT transfer where host controller is the one responsible for PID toggle. Though there is not many control transfer that has large payload > 64. We can keep this in mind, should it be an issue in the future.

Reply:
> well i'm still extremely confused by the code... at least a comment is needed... i'm still not clear why you toggle the packet every buffer (and indeed whether that is a 8 byte buffer or not) when the total packet length is odd. what if there are 4 packets?

I have explained it very clearly as above. PID flipped every packet per USB specs, 4 wont flip it